kernel/sched.c

```diff
@@ -1347,3 +1372,49 @@ int k_thread_cpu_mask_disable(k_tid_t thread, int cpu)
 }
 
 #endif /* CONFIG_SCHED_CPU_MASK */
+
+int z_impl_k_thread_join(struct k_thread *thread, s32_t timeout)
+{
+	k_spinlock_key_t key;
+	int ret;
+
+	__ASSERT(((arch_is_in_isr() == false) || (timeout == K_NO_WAIT)), "");
+
+	key = k_spin_lock(&sched_spinlock);
+
+	if ((thread->base.thread_state & _THREAD_DEAD) != 0) {
+		ret = 0;
+		goto out;
+	}
+
+	if (timeout == K_NO_WAIT) {
+		ret = -EBUSY;
+		goto out;
+	}
+
+	if (thread->base.pended_on == &_current->base.join_waiters) {
+		ret = -EDEADLK;
+		goto out;
+	}
+
+#if defined(CONFIG_TIMESLICING) && defined(CONFIG_SWAP_NONATOMIC)
+	pending_current = _current;
+#endif
+	add_to_waitq_locked(_current, &thread->base.join_waiters);
+	add_thread_timeout_ms(_current, timeout);
+
+	return z_swap(&sched_spinlock, key);
+out:
+	k_spin_unlock(&sched_spinlock, key);
+	return ret;
+}
+
+#ifdef CONFIG_USERSPACE
+int z_vrfy_k_thread_join(struct k_thread *thread, s32_t timeout)
+{
+	Z_OOPS(Z_SYSCALL_OBJ(thread, K_OBJ_THREAD));
```

Comment:
this (and also k_thread_abort()) needs fixing.

if the thread object is uninitialized (which means it never ran, or already ran and exited) should just return 0.

the initialization bit is pulling double-duty for thread objects. don't see away around this without splitting k_thread_create() into separate calls to initialize the object to a sane state, and spawn the thread.